java/arcs/core/util/Parser.kt

```diff
@@ -0,0 +1,278 @@
+ * null.
+ */
+class Optional<T>(val parser: Parser<T>) : Parser<T?>() {
+    override fun invoke(string: String): ParseResult<T?> = parser(string).orElse {
+        Success<T?>(null, string)
+    }
+}
+
+/**
+ * A parser that combines two parsers in sequence, if either one fails, the parse fails,
+ * otherwise the combined results are returned as a [Pair<T, S].
+ */
+class Seq<T, S>(val left: Parser<T>, val right: Parser<S>) : Parser<Pair<T, S>>() {
+    override fun invoke(string: String) = left(string).flatMap { v1, r ->
+        right(r).flatMap { v2, r2 ->
+            Success(Pair(v1, v2), r2)
+        }
+    }
+}
+
+/** A parser that combines two parsers by returning the value of the first one that succeeds. */
+class Par<T>(val left: Parser<T>, val right: Parser<T>) : Parser<T>() {
+    override fun invoke(string: String): ParseResult<T> = left(string).orElse {
+        right(string)
+    }
+}
+
+/**
+ * A parser that invokes another parser zero or more times until it fails, returning a list
+ * of success values. If no parse succeeds, it returns an empty list.
+ */
+class Many<T>(val parser: Parser<T>) : Parser<List<T>>() {
+    override fun invoke(string: String): ParseResult<List<T>> {
+        val result = mutableListOf<T>()
+        var parseResult = parser(string)
+        var lastSuccess = parseResult
+        while (parseResult is Success<T>) {
+            result.add(parseResult.value)
+            lastSuccess = parseResult
+            parseResult = parser(parseResult.rest)
+        }
+        return Success(result, (lastSuccess as? Success<T>)?.rest ?: string)
+    }
+}
+
+/** A parser which converts the return value of a parser into another value. */
+class Transform<T, R>(val parser: Parser<T>, val transform: (T) -> R) : Parser<R>() {
+    override fun invoke(string: String): ParseResult<R> = parser(string).flatMap { v, r ->
+        Success(transform(v), r)
+    }
+}
+
+/** A parser used to refer to parsers that haven't been constructed yet. */
+class LazyParser<T>(val parser: () -> Parser<T>) : Parser<T>() {
+    override fun invoke(string: String): ParseResult<T> = parser()(string)
+}
+
+/** A parser that represents three parsers in sequence yielding a [Triple]. */
+class TriSeq<T, S, R>(val left: Seq<T, S>, val right: Parser<R>) : Parser<Triple<T, S, R>>() {
+    override fun invoke(string: String): ParseResult<Triple<T, S, R>> =
+        left(string).flatMap { v1, r1 ->
+            right(r1).flatMap { v2, r2 ->
+                Success(Triple(v1.first, v1.second, v2), r2)
+            }
+        }
+}
+
+/** A parser which omits its output from the result type. */
+class IgnoreParser<T>(val parser: Parser<T>) : Parser<T>() {
+    override fun invoke(string: String): ParseResult<T> = parser(string)
+}
+
+/** Combines two parsers via addition operator as [Seq] combinator. */
+inline operator fun <reified T, reified S> Parser<T>.plus(other: Parser<S>) = Seq(this, other)
```

Comment:
Do you _need_ these type arguments to be `reified` ? I doesn't seem to me like you do.

Reply:
No, it's a left over from some tricks I was trying to do with inferring Tuples many levels down and the + operator. Turns out not to be needed once I introduced the TriSeq. I was trying to make it work with just Seq<> and a single plus operator.